fendermint/vm/interpreter/src/fvm/genesis.rs

```diff
@@ -80,92 +91,296 @@ where
+                TokenAmount::zero(),
+                None,
+            )
+            .context("failed to create burnt funds actor")?;
 
         // A placeholder for the reward actor, beause I don't think
         // using the one in the builtin actors library would be appropriate.
         // This effectively burns the miner rewards. Better than panicking.
-        state.create_actor(
-            account::ACCOUNT_ACTOR_CODE_ID,
-            reward::REWARD_ACTOR_ID,
-            &account::State {
-                address: reward::REWARD_ACTOR_ADDR,
-            },
-            TokenAmount::zero(),
-            None,
-        )?;
-
-        // Create accounts
+        state
+            .create_actor(
+                account::ACCOUNT_ACTOR_CODE_ID,
+                reward::REWARD_ACTOR_ID,
+                &account::State {
+                    address: reward::REWARD_ACTOR_ADDR,
+                },
+                TokenAmount::zero(),
+                None,
+            )
+            .context("failed to create reward actor")?;
+
+        // STAGE 2: Create non-builtin accounts which do not have a fixed ID.
+
+        // The next ID is going to be _after_ the accounts, which have already been assigned an ID by the `Init` actor.
+        // The reason we aren't using the `init_state.next_id` is because that already accounted for the multisig accounts.
         let mut next_id = init::FIRST_NON_SINGLETON_ADDR + addr_to_id.len() as u64;
+
         for a in genesis.accounts {
             let balance = a.balance;
             match a.meta {
                 ActorMeta::Account(acct) => {
-                    state.create_account_actor(acct, balance, &addr_to_id)?;
+                    state
+                        .create_account_actor(acct, balance, &addr_to_id)
+                        .context("failed to create account actor")?;
                 }
                 ActorMeta::Multisig(ms) => {
-                    state.create_multisig_actor(ms, balance, &addr_to_id, next_id)?;
+                    state
+                        .create_multisig_actor(ms, balance, &addr_to_id, next_id)
+                        .context("failed to create multisig actor")?;
                     next_id += 1;
                 }
             }
         }
 
-        Ok((state, output))
+        // STAGE 3: Initialize the FVM and create built-in FEVM actors.
+
+        state
+            .init_exec_state(
+                out.timestamp,
+                out.network_version,
+                out.base_fee.clone(),
+                out.circ_supply.clone(),
+                out.chain_id.into(),
+            )
+            .context("failed to init exec state")?;
+
+        let mut deployer = ContractDeployer::<DB>::new(&self.contracts);
+
+        // IPC libraries.
+        for (lib_src, lib_name) in eth_libs {
+            deployer.deploy_library(&mut state, &mut next_id, lib_src, &lib_name)?;
+        }
+
+        // IPC Gateway actor.
+        let gateway_addr = {
```

Comment:
This might run into a contract size too big erorr. Also we are migrating to diamond pattern that resolves the size issue.

Reply:
Can you elaborate on what exactly will get the error? The way it is deployed, or the `gateway_addr` in particular?

I actually had to create a branch on `ipc-solidity-actors` which [I am still checking out](https://github.com/consensus-shipyard/fendermint/blob/c11bd2ee0d08c14a72728f6208f33087b1c22c10/Makefile#L103) in this PR which comments out some methods, but I saw @adlrocha also has a [version of gateway.rs](https://github.com/consensus-shipyard/fendermint/pull/160/files#diff-923a0205e0b6a0e34d6f2e17ab68481b0fe6cbb1e78586d6d20202e851d48ea5) with the extra methods included in his PR, although I am not sure if that has been a clean build in the sense that it if I used `make ipc-actors` in _this_ repo, it would exit with a failure and stop the rest of the build, but if the artifacts already existed, then maybe Rust code generation is fine, and only deployment would stop. In any case, this PR uses the a smaller version. 

I would assume the library linking will be the same with the diamond pattern as well, no?

Reply:
Thanks for the link, it looks way more complicated now :flushed: